Add the first round of light and mass profiles

CKrawczyk opened the pull request:
See the new README for a full list of all profiles added.

Also, set up Flit to be the Python build system to turn this repo into a simple Python package.

CKrawczyk committed:
Turn the repository into a python package

I am using Flit as the python build system to make the repository into a light weight python package that can be `pip install`ed locally.

CKrawczyk committed:
Add new light profile

Only one new profile, a vectorized version of the gaussian light profile.

CKrawczyk committed:
Add new mass profiles

These are the various profiles the UoP has developed for use in recent (~2025) publications.  They included:

- GaussianKappa
- GaussianEllipseKappa
- MultiGaussianEllipseKappa
- MGE
- NFW
- TNFW

CKrawczyk committed:
Update the readme to list all the profiles

WolfgangEnzi commented on README.md:
```diff
@@ -1,2 +1,17 @@
+
+## Light Profiles
+- `multi_gaussian_light.MultiGaussian`: Sum of multiple (vectorized) elliptical Gaussian profiles
+
```
Add Multi Gaussian Ellipse to this list !

WolfgangEnzi commented on jax_lensing_profiles/LightModel/Profiles/multi_gaussian_light.py:
```diff
@@ -0,0 +1,144 @@
+# Defines a gaussian profile
```
Change to Multi Gaussian light profile.  Update Copyright to 2025.

CKrawczyk committed:
Add __init__ files

These are need to make the Python packaging work.

WolfgangEnzi commented on jax_lensing_profiles/MassModel/Profiles/NFW.py:
```diff
@@ -0,0 +1,227 @@
+from jax.tree_util import Partial as partial
+
+# References:
+# https://arxiv.org/pdf/0705.0682.pdf
```
One could comment on the fact that according to equation (2) of https://arxiv.org/pdf/0705.0682.pdf, we assume n=1.

WolfgangEnzi commented on jax_lensing_profiles/LightModel/Profiles/multi_gaussian_light.py:
```diff
@@ -0,0 +1,144 @@
+        :param amp: amplitude, such that 2D integral leads to this value
+        :param sigma: sigma of Gaussian in each direction
+        :param center_x: center of profile
+        :param center_y: center of profile
```
I think these should be lists.

WolfgangEnzi commented on jax_lensing_profiles/LightModel/Profiles/multi_gaussian_light.py:
```diff
@@ -0,0 +1,144 @@
+        :param amp: amplitude, such that 2D integral leads to this value
+        :param sigma: sigma of Gaussian in each direction
+        :param center_x: center of profile
+        :param center_y: center of profile
```
These will be lists. Also since this only depends on the amplitudes, it could be worth removing the other arguments.

WolfgangEnzi commented on jax_lensing_profiles/LightModel/Profiles/multi_gaussian_light.py:
```diff
@@ -0,0 +1,144 @@
+
+        :param r: 3d distance from center of profile
+        :param amp: amplitude, such that 2D integral leads to this value
+        :param sigma: sigma of Gaussian in each direction
```
Again lists. This also seems to assume that the Gaussian components share a center. This should be made clear in the documentation.

WolfgangEnzi commented on jax_lensing_profiles/LightModel/Profiles/multi_gaussian_light.py:
```diff
@@ -0,0 +1,144 @@
+        :param e1: eccentricity modulus
+        :param e2: eccentricity modulus
+        :param center_x: center of profile
+        :param center_y: center of profile
```
Again Lists.

WolfgangEnzi commented on jax_lensing_profiles/LightModel/Profiles/multi_gaussian_light.py:
```diff
@@ -0,0 +1,144 @@
+        :param e1: eccentricity modulus
+        :param e2: eccentricity modulus
+        :param center_x: center of profile
+        :param center_y: center of profile
```
Again Lists.

CKrawczyk committed:
Make __init__ files do useful things

Import all the profiles into a useful structure.  Also register all the new profiles with Herculens on import.

WolfgangEnzi commented on jax_lensing_profiles/LightModel/Profiles/multi_gaussian_light.py:
```diff
@@ -0,0 +1,144 @@
+        :param amp: amplitude, such that 2D integral leads to this value
+        :param sigma: sigma of Gaussian in each direction
+        :param e1: eccentricity modulus
+        :param e2: eccentricity modulus
```
Again lists.   Maybe also make a comment on eccentricity and ellipticity nomenclature!

WolfgangEnzi commented on jax_lensing_profiles/MassModel/Profiles/MGE_jax.py:
```diff
@@ -0,0 +1,257 @@
+__author__ = "CKrawczyk"
```
Add description like:

# Defines the MGE expansion according to [References]. And comment on copyrights.

WolfgangEnzi commented on jax_lensing_profiles/MassModel/Profiles/MGE_jax.py:
```diff
@@ -0,0 +1,257 @@
+
+
+class MGE(object):
+    def __init__(
```
Could be very helpful to add docstring for this class!

WolfgangEnzi commented on jax_lensing_profiles/MassModel/Profiles/MGE_jax.py:
```diff
@@ -0,0 +1,257 @@
+        self.sigma_start_mult = sigma_start_mult
+        self.sigma_end_mult = sigma_end_mult
+        
+        # eq. 6 for fixed nodes and weights
```
Reference is missing for this equation!

WolfgangEnzi replied:
https://academic.oup.com/mnras/article/488/1/1387/5526256

CKrawczyk committed:
Update readme with more info

Now including `Installation` and `Usage` sections.

CKrawczyk committed:
First pass at file headers

New profiles have a docstring on the file that gives the references used.

CKrawczyk committed:
Use the MGE profile to define a Sersic profile

This is a mass profile that has and elliptical Sersic profile for the convergence.

WolfgangEnzi commented on jax_lensing_profiles/MassModel/Profiles/NFW.py:
```diff
@@ -0,0 +1,227 @@
+# https://arxiv.org/pdf/1101.0650.pdf
+# https://arxiv.org/pdf/astro-ph/0102341
+
+# ============================================================================================
+# Base class for NFW and tNFW
+# ============================================================================================
```
Comments do not follow the style of the other files.

WolfgangEnzi commented on jax_lensing_profiles/MassModel/Profiles/NFW.py:
```diff
@@ -0,0 +1,227 @@
+        return xr, tau
+
+    @staticmethod
+    def _psi_truncated_nfw(x, tau):
```
Equation A.14 of https://arxiv.org/pdf/0705.0682

CKrawczyk committed:
Add elliptical NFW kappa profile

Use MGEs to define an elliptical NFW profile.

WolfgangEnzi commented on jax_lensing_profiles/MassModel/Profiles/NFW.py:
```diff
@@ -0,0 +1,227 @@
+
+class NFW(NFWBase):
+    '''
+    this class contains function to evaluate the NFW derivatives
```
Maybe change to the "NFW lensing potential derivatives".

CKrawczyk committed:
Update MGE derived classes

These need to be subclasses of the MGE class to work as expected.  The only change is the `__init__` method is overridden to call the MGE `__init__` with the correct function for it to wrap.

WolfgangEnzi commented on jax_lensing_profiles/MassModel/Profiles/NFW.py:
```diff
@@ -0,0 +1,227 @@
+        )(x, y)
+        return h[..., 0, 0], h[..., 1, 1], h[..., 0, 1]
+
+# ============================================================================================
+# Class for NFW
+# ============================================================================================
```
Again not the same comment style as in previous files.

CKrawczyk committed:
Update the readme with an MGE example

How to use the MGE profile class is a bit non-intuitive (e.g. you need to subclass it rather than just call it).  The readme now provides and example for how to use it.

CKrawczyk committed:
Rename file for consistency

CKrawczyk committed:
Add new MGE based profiles to the readme

WolfgangEnzi commented on jax_lensing_profiles/MassModel/Profiles/NFW.py:
```diff
@@ -0,0 +1,227 @@
+        )
+
+
+# ============================================================================================
+# Class for tNFW
+# ============================================================================================
```
Style of the comment.

WolfgangEnzi commented on jax_lensing_profiles/MassModel/Profiles/NFW.py:
```diff
@@ -0,0 +1,227 @@
+        xr = NFW.center_and_scale(x, y, rs, center_x, center_y)
+        # Fx = NFW.F(xr)
+        # Fx = F(xr)
+        # psi = jnp.log(0.5 * xr)**2 - Fx * Fx * (1 - xr**2)
```
It could be helpful to outline the connection between F(x) and H(x) more clearly in the f_function_jax file later on!

WolfgangEnzi replied:
https://arxiv.org/pdf/astro-ph/0102341 This might be one of the references you want to provide here.

WolfgangEnzi replied:
Alternatively ( https://arxiv.org/pdf/2310.03077 which refers to https://arxiv.org/abs/astro-ph/0201501v2)

WolfgangEnzi commented on jax_lensing_profiles/Utility/f_function_jax.py:
```diff
@@ -0,0 +1,152 @@
+import jax.numpy as jnp
```
Comment at the top describing what is defined.
The references in NFW.py might be useful to be added here as well.

WolfgangEnzi replied:
Potentially also add a description that shows that the definitions that are provided in the references are equivalent.

WolfgangEnzi commented on jax_lensing_profiles/Utility/f_function_jax.py:
```diff
@@ -0,0 +1,152 @@
+#     # tangent_out = (x * masked_out - 1/x) / (1 - x**2)
+
+#     tangent_out = jnp.empty_like(x)
+#     tangent_out = jnp.where(
+#         near_one,
+#         -2/3 + (14/15) * (x - 1),
+#         (x * primal_out - 1/x) / (1 - x**2)
+#     )
+#     return primal_out, x_dot * tangent_out
+
+
+# @custom_jvp
+# def H(x):
+#     xm1 = x - 1
+#     output = jnp.empty_like(x)
+#     output = jnp.where(
+#         x <= 0.99999,
+#         -jnp.arccosh(1 / x)**2,
+#         output
+#     )
+#     output = jnp.where(
+#         x >= 1.00001,
+#         jnp.arccos(1 / x)**2,
+#         output
+#     )
+#     output = jnp.where(
+#         (x > 0.99999) & (x < 1.00001),
+#         2 * xm1 - (5/3) * xm1**2 + (64/45) * xm1**3 - (26/21) * xm1**4,
+#         output
+#     )
+#     return output
+
+
+# @H.defjvp
+# def H_jvp(primals, tangents):
+#     x, = primals
+#     x_dot, = tangents
+#     primal_out = H(x)
+#     tangent_out = 2 * F(x) / x
+#     return primal_out, x_dot * tangent_out
```
Alternative definitions can be removed from current version.

WolfgangEnzi commented on jax_lensing_profiles/Utility/f_function_jax.py:
```diff
@@ -0,0 +1,152 @@
+    return ((t2 - t1) / somr2).real
+
+
+# def F(r):
+#     return 0.5 * r * grad_H(r)
+
```
Can be removed from current version.

WolfgangEnzi commented on jax_lensing_profiles/Utility/f_function_jax.py:
```diff
@@ -0,0 +1,152 @@
+    return -((t2 - t1)**2).real
+
+
+# grad_H = jnp.vectorize(
+#     grad(H),
+#     signature='()->()'
+# )
```
Can be removed.

WolfgangEnzi commented on pyproject.toml:
```diff
@@ -0,0 +1,17 @@
+
+[project]
+name = "jax_lensing_profiles"
+authors = [{name = "Coleman Krawczyk", email = "coleman.krawczyk@port.ac.uk"}]
```
Add other authors names.

WolfgangEnzi commented on jax_lensing_profiles/MassModel/Profiles/NFW.py:
```diff
@@ -0,0 +1,227 @@
+            rt=rt,
+            center_x=center_x,
+            center_y=center_y
+        )
```
We could add to this files other helpful functions, e.g. the total mass of the truncated NFW profile.

WolfgangEnzi commented on jax_lensing_profiles/Utility/faddeeva_function.py:
```diff
@@ -0,0 +1,101 @@
+import jax.numpy as jnp
```
Add comment at the top that describes what this python script defines.
Add references for this (I assume it was https://academic.oup.com/mnras/article/488/1/1387/5526256 and /or https://dl.acm.org/doi/10.1145/3119904 which you already mention below.)

WolfgangEnzi commented on jax_lensing_profiles/Utility/faddeeva_function.py:
```diff
@@ -0,0 +1,101 @@
+
+@custom_jvp
+def w_f(z):
+    """Compute the Faddeeva function :math:`w_{\\mathrm F}(z)` using the
```
I might be wrong, but I think I remember that this expansion goes to a higher order  expansion than is usually used. It might be good to comment on why this higher order is included if this is the case.

WolfgangEnzi commented on jax_lensing_profiles/Utility/basic_quad_jax.py:
```diff
@@ -0,0 +1,40 @@
+import jax
```
# Describe what this defined in this script.

WolfgangEnzi commented on jax_lensing_profiles/MassModel/Profiles/multi_gaussian_ellipse_kappa_jax.py:
```diff
@@ -0,0 +1,232 @@
+        ).sum()
+
+    def function(self, x, y, amp, sigma, e1, e2, center_x, center_y):
+        part = partial(
```
Add docstring to function, as it takes lists of amps, sigmas, e1, e2.

WolfgangEnzi commented on jax_lensing_profiles/MassModel/Profiles/multi_gaussian_ellipse_kappa_jax.py:
```diff
@@ -0,0 +1,232 @@
+        return jnp.stack([f_x, f_y])
+    
+    def derivatives(self, x, y, amp, sigma, e1, e2, center_x, center_y):
+        part = partial(
```
Add docstring.

WolfgangEnzi commented on jax_lensing_profiles/MassModel/Profiles/multi_gaussian_ellipse_kappa_jax.py:
```diff
@@ -0,0 +1,232 @@
+        )(x, y, amp, sigma, e1, e2, center_x=center_x, center_y=center_y))
+
+    def hessian(self, x, y, amp, sigma, e1, e2, center_x, center_y):
+        part = partial(
```
Add docstring.

WolfgangEnzi commented on jax_lensing_profiles/MassModel/Profiles/multi_gaussian_ellipse_kappa_jax.py:
Name of the file could be made more uniform w.r.t. to the light models.
Either call this file multi_gaussian_kappa.py, or the light distribution file multi_gaussian_ellipse_light.py.

CKrawczyk replied:
The light file has both the circular and elliptical versions inside, for the mass they are in different files.  Although I might drop the `_jax` from the file names.

WolfgangEnzi commented on jax_lensing_profiles/MassModel/Profiles/gaussian_kappa_jax.py:
```diff
@@ -0,0 +1,127 @@
+import jax.numpy as jnp
```
Docstring describing what is defined in this file.

WolfgangEnzi commented on jax_lensing_profiles/MassModel/Profiles/gaussian_kappa_jax.py:
```diff
@@ -0,0 +1,127 @@
+        return alpha
+
+    def function(self, x, y, amp, sigma, center_x=0, center_y=0):
+        """Returns Gaussian."""
```
Returns the lensing potential created by a Gaussian convergence.

WolfgangEnzi commented on jax_lensing_profiles/MassModel/Profiles/gaussian_kappa_jax.py:
```diff
@@ -0,0 +1,127 @@
+        return integral_term * amp2d
+
+    def derivatives(self, x, y, amp, sigma, center_x=0, center_y=0):
+        xc, yc, R = GaussianKappa.center(x, y, center_x, center_y)
```
Returns the deflection angles created by a Gaussian convergence.

WolfgangEnzi commented on jax_lensing_profiles/MassModel/Profiles/gaussian_kappa_jax.py:
```diff
@@ -0,0 +1,127 @@
+        return 1.0 / R**2 * (-1 + (1 + 2 * c * R**2) * jnp.exp(-c * R**2)) * A
+
+    def hessian(self, x, y, amp, sigma, center_x=0, center_y=0):
+        xc, yc, R = GaussianKappa.center(x, y, center_x, center_y)
```
Add docstring.

WolfgangEnzi commented on jax_lensing_profiles/MassModel/Profiles/MGE_jax.py:
```diff
@@ -0,0 +1,257 @@
+        )
+
+    def function(self, x, y, **kwargs):
+        # vectorize over x and y
```
Add docstring, in particular what the user has to add in kwargs.

Somthing like: Returns the lensing potential of a kappa that is parameterized as the sum of multiple Gaussian distributions.

WolfgangEnzi commented on jax_lensing_profiles/MassModel/Profiles/MGE_jax.py:
```diff
@@ -0,0 +1,257 @@
+        return jnp.stack([f_x, f_y])
+    
+    def derivatives(self, x, y, e1, e2, center_x=0, center_y=0, **kwargs):
+        part = partial(
```
Add docstring, in particular what needs to be added as kwargs.

WolfgangEnzi commented on jax_lensing_profiles/MassModel/Profiles/MGE_jax.py:
```diff
@@ -0,0 +1,257 @@
+        )(x, y, e1, e2, center_x=center_x, center_y=center_y, **kwargs))
+    
+    def hessian(self, x, y, e1, e2, center_x=0, center_y=0, **kwargs):
+        part = partial(
```
Add docstring.

WolfgangEnzi commented on jax_lensing_profiles/MassModel/Profiles/gaussian_ellipse_kappa_jax.py:
```diff
@@ -0,0 +1,223 @@
+        return x_, y_, cos_phi, sin_phi
+
+    def function(self, x, y, amp, sigma, e1, e2, center_x=0, center_y=0):
+        phi_g, q = param_util.ellipticity2phi_q(e1, e2)
```
Add docstring.

WolfgangEnzi commented on jax_lensing_profiles/MassModel/Profiles/gaussian_ellipse_kappa_jax.py:
```diff
@@ -0,0 +1,223 @@
+        )
+
+    def derivatives(self, x, y, amp, sigma, e1, e2, center_x=0, center_y=0):
+        phi_g, q = param_util.ellipticity2phi_q(e1, e2)
```
Add docstring.

WolfgangEnzi commented on jax_lensing_profiles/MassModel/Profiles/gaussian_ellipse_kappa_jax.py:
```diff
@@ -0,0 +1,223 @@
+        return f_x, f_y
+
+    def hessian(self, x, y, amp, sigma, e1, e2, center_x=0, center_y=0):
+        phi_g, q = param_util.ellipticity2phi_q(e1, e2)
```
Add docstring.

CKrawczyk committed:
Add multi-gaussian docstrings

Also removed the 3D light flux as I am unsure how the centers should be handled

CKrawczyk committed:
MGE family of docstrings

Making sure `function`, `derivatives`, and `hessian` get documentation.

CKrawczyk committed:
All the docstrings

Full pass over the remaining files adding in consistent docstrings for all relevant functions and methods.

Docstrings are also included for inherited methods of subclasses.

CKrawczyk committed:
Rename files to be more consistent

don't need `_jax` at the end of everything.

CKrawczyk committed:
Add new code author
